feat: enable route protection

OkelleyDevelopment opened the pull request:
### Summary 

- Add `protection` middleware to verify the JWTs to ensure they are present and current.
- Enabled on the posts router
- Updated to have a sample test, can be refactored later once the posts endpoint is more developed.

OkelleyDevelopment committed:
feat: Add jwt validation middleware

OkelleyDevelopment committed:
feat: secure posts router

Add the protect middleware (to verify a valid JWT) to the posts router.

AceMouty commented on src/api/middleware/verify_json.ts:
```diff
@@ -12,7 +12,7 @@ import StatusCodes from "../utils/status_codes";
-    const username = req.body.username
+    const username = req.body.username;
     if (!username) {
       res.status(StatusCodes.UNAUTHORIZED).send({
```
Just realized use of `send` instead of `json`
replace instances of `.send` with `.json` in this file, we should be replying to the requesting client with JSON

https://expressjs.com/en/4x/api.html#res.json

OkelleyDevelopment replied:
Ah good catch on that one :raised_hands:

AceMouty commented on src/api/middleware/verify_json.ts:
```diff
@@ -30,7 +30,7 @@ export const verify_username = expressAsyncHandler(
-    const password = req.body.password
+    const password = req.body.password;
     if (!password) {
       res.status(StatusCodes.UNAUTHORIZED).send({
```
Replace instances of `send` with `json`

AceMouty commented on src/api/middleware/verify_json.ts:
```diff
@@ -48,7 +48,7 @@ export const verify_password = expressAsyncHandler(
-    const roleId = req.body.roleId
+    const roleId = req.body.roleId;
     if (!roleId) {
       res.status(StatusCodes.UNAUTHORIZED).send({
```
Replace instances of `send` with `json`

AceMouty commented on src/api/middleware/verify_jwt.ts:
```diff
@@ -0,0 +1,51 @@
+      } catch (error) {
+        res
+          .status(StatusCodes.UNAUTHORIZED)
+          .send({ message: "Not Authorized" });
```
Replace instances of `send` with `json`

AceMouty commented on src/api/middleware/__tests__/verify_jwt.test.ts:
```diff
@@ -0,0 +1,11 @@
+
+import StatusCodes from "../../utils/status_codes";
+
+describe("Testing a sample protected route", () => {
```
add tests checking the data is coming back in JSON format and also that the shape of the data is correct eg. do we have keys on the response object that we expect to have.

OkelleyDevelopment committed:
fix: change .send() to .json()

OkelleyDevelopment committed:
feat: Add new unit test

This unit test checks for the content type to be json and with
a body that has the key 'message'.
